client/gutenberg/editor/init.js

```diff
@@ -34,26 +34,24 @@ const addResetToRegistry = registry => {
 		window.gutenbergState = () => mapValues( registry.stores, ( { store } ) => store.getState() );
 	}
 
-	const resettableStores = [ 'core/editor', 'core/notices' ];
-
 	const stores = [];
 	return {
 		registerStore( namespace, options ) {
```

Comment:
@aduth @jsnajdr
This part here that extends all the stores with a `reset` method, is not applied to `core/data` (aka the store in the `data` package) as well.
It's easily testable by adding a `console.log( namespace )` here, and observe that it "traverses" all expected stores, except `core/data`.
I'm not sure why, and I always get lost in the logic behind the Gutenberg data module.

In other words, this function here allows me to reset `core` (aka the store in the `core-data` package) but not `core/data`.

Reply:
The reason why the `core/data` ignores the call to `reset()` is that it's created and [registered very early](https://github.com/WordPress/gutenberg/blob/e1092c0d0b75fe53ab57bc6c4cc9e32cb2e74e40/packages/data/src/registry.js#L169-L172) as part of the `createRegistry` function.

At that time, the `registerStore` function is not yet extended with the `use( addResetToRegistry )` plugin and the registered reducer is not wrapped with the HoR that reacts to the `GUTENLYPSO_RESET` action.

I don't think that being able to fully reset the `core/data` store is even wanted. The `core/data` store is a bit special: it maintains the resolution flag for all other stores' selectors. We only ever want to reset the resolution flags for one store at a time, not for all of them.

Fortunately, some support for resetting `core/data` state was added in #12877 and we can use it for our purposes. Inspired by the [resolvers cache middleware](https://github.com/WordPress/gutenberg/blob/e1092c0d0b75fe53ab57bc6c4cc9e32cb2e74e40/packages/data/src/resolvers-cache-middleware.js#L14-L34), we can implement a function that resets all resolution state for one store:
```js
function resetStoreResolutions( registry, reducerKey ) {
  const resolvers = registry.select( 'core/data' ).getCachedResolvers( reducerKey );
  Object.entries( resolvers ).forEach( ( [ selectorName, resolversByArgs ] ) => {
    resolversByArgs.forEach( ( _, args ) => {
      registry.dispatch( 'core/data' ).invalidateResolution( reducerKey, selectorName, args );
    } );
  } );
}
```

Then we can add a call to this new function to the `reset` function:
```js
reset() {
  resettableStores.forEach( reducerKey => {
    registry.stores[ reducerKey ].dispatch( { type: 'GUTENLYPSO_RESET' } );
    resetStoreResolutions( registry, reducerKey );
  } );
}
```

This should reset everything on site change. There is just one race condition bug that remains: if a request for entities from site ID=1 is in flight, and we switch to site ID=2, then the response from site ID=2 will be received into store with site ID=2!

The cache invalidation code in #12877 probably has a similar bug: it won't invalidate resolution for selector whose request is in flight. There can be "get a list" and "add to list" requests happening at the same time, and if their "issue request from client", "perform request on server" and "receive response on client" are performed in a certain order, the resulting state will miss the just-added entity.

Reply:
Created an issue for tracking at https://github.com/WordPress/gutenberg/issues/12984